### PR TITLE
Add a fn `listen_addresses()` to NetworkService

### PR DESCRIPTION
**Pull request by cecton**
# Use case

I'm writing a test crate for polkadot and a simple test that ensures blocks are being built. For that I need to start a few nodes on random ports. To do that I pass the port 0 to get a random port but I need to know what port has been assigned for each node.

# Proposed solution

The only way I found would be to call `listen_addresses` on the NetworkWorker but since I only have a NetworkService, I added a function `listen_addresses` to NetworkService which passes the message to the NetworkWorker.

### CODE REVIEW

**Comment by tomaka** [comment]
For testing, note that the network can be put into a "memory-only" mode: https://github.com/paritytech/substrate/blob/d5336db1b010c33f30ad7d1644038900539c264e/client/network/src/config.rs#L535

The idea is to generate `u64` port numbers ahead of time (using `rand::random::<u64>()`), then you pass addresses like `/memory/1234` as listening addresses.

Example usage: https://github.com/paritytech/substrate/blob/adb5acbe7e0216a641d55056d4f016af9e8cf0c7/client/network/src/protocol/generic_proto/tests.rs#L35-L112

Beyond the fact that it solves the problem that you're having, it also solves the problem that using TCP sockets might also lead to running out of ports if you have several heavy tests running at the same time.

File: client/network/src/service.rs [hunk 1]
```diff
@@ -747,6 +747,15 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
 			.unbounded_send(ServiceToWorkerMsg::UpdateChain);
 	}
 
+	/// Returns a stream containing the listen addresses
```
Comment (tomaka) [inline client/network/src/service.rs:750]:
```suggestion
	/// Returns the list of addresses we are currently listening on.
	///
	/// Please be aware that this list might change over time.
```

**Comment by cecton** [comment]
This is AWESOME!! Thanks! ... :thinking: this is strange. Now my function listen_addresses return an empty Vec. Shouldn't it have the memory address?

**Comment by tomaka** [comment]
> Shouldn't it have the memory address?

Oh, this might be a bug in rust-libp2p. The list of listened addresses is reported by the low-level transport code (such as the TCP code), and I think we don't do that for the memory transport.

**Comment by tomaka** [comment]
Nevermind, we do it: https://github.com/libp2p/rust-libp2p/blob/168b00ed3e8999904e075469eb867d692e04487f/core/src/transport/memory.rs#L179-L182

Another possibility is that because opening listeners is actually asynchronous in libp2p, you're processing the nodes in such a way that you get the list of listened addresses before the first listener has opened.

**Comment by cecton** [comment]
I did some experiment but I have a new issue: it seems my nodes are not connected for some reason (it says "0 peers")... I haven't investigated yet

```
2020-06-18 18:08:46 Can't listen on /memory/18035446117391323389 because: MultiaddrNotSupported("/memory/18035446117391323389")
2020-06-18 18:08:51 💤 Idle (0 peers), best: #0 (0x854f…eb5f), finalized #0 (0x854f…eb5f), ⬇ 0 ⬆ 0
2020-06-18 18:08:54 💤 Idle (0 peers), best: #0 (0x854f…eb5f), finalized #0 (0x854f…eb5f), ⬇ 0 ⬆ 0
2020-06-18 18:08:56 💤 Idle (0 peers), best: #0 (0x854f…eb5f), finalized #0 (0x854f…eb5f), ⬇ 0 ⬆ 0
```

Also: after tweaking with all of this I have the feeling that `listen_addresses` would still be good to have because it reduces the code on my side (I won't have to create a MultiaddrWithPeerId struct to determine the boot_node).

**Comment by cecton** [comment]
Thanks for the help @tomaka I had to set the TransportConfig of the `sc-service::Configuration` to `MemoryOnly` to make it work. I might open up a pr to improve the error message on that.

I'm closing this PR for now as it would be unnecessary code to maintain and I will post my solution soon when it is merged.

**Pull request closed without merging**